go/pkg/pass1/print-service.1.md
# print-service 1 "" Netspoc "User Manual"

# NAME

print-service - Show rules of netspoc service definitions

# SYNOPSIS

print-service [options] FILE|DIR [SERVICE-NAME ...]

# DESCRIPTION

This program prints expanded rules of given Netspoc service definitions.
If no service name is given, all services are printed.

Output format is

with option `-i` (default)
:   `service-name:permit|deny src-ip dst-ip protocol-description`

with option `-n`
:   `service-name:permit|deny src-name dst-name protocol-description`

with option `-i -n`
:   `service-name:permit|deny src-ip src-name dst-ip dst-name protocol-description`

# OPTIONS

**-n**, **--name**
:   Show name of elements.

**-i**, **--ip**
:   Show IP address of elements.

**--nat** name
:   Uses network:name as reference when resolving IP address in a NAT environment.

**-q**, **--quiet**
:   Don't print progress messages.

**-h**, **--help**
:   Print a brief help message and exit.

# COPYRIGHT AND DISCLAIMER

(c) 2025 by Heinz Knutzen, heinz.knutzen@googlemail.com

This program is part of Netspoc, a Network Security Policy Compiler.
http://hknutzen.github.com/Netspoc

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY OR FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License along
with this program; if not, write to the Free Software Foundation, Inc.,
51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.


Changes
{{$NEXT}}

 - export-netvis: export Dualstack-Address of network
   in field Address2 of node.

6.079     2025-03-17 12:53:54+01:00 Europe/Berlin

 - If service has some owner with name "DA_*",
   this service is marked as visible for all owners with name "DA_*"
   in exported data for Netspoc-Web.
 - Command 'print-path' now accepts dual stack objects.

6.078     2025-02-24 14:42:54+01:00 Europe/Berlin

 - New commandline programs export-netvis and print-path have been added.

    These tools output information as JSON to be used in
    a graph visualization.

    export-netvis creates a list of all networks and routers
    of the given topology with information useful for visualizing
    the topology.

    print-path analyzes the given topology and creates a list of
    names of network nodes (routers and networks) between source and
    destination passed via command line.
    To create the list, print-path uses the same path-walk algorithm
    as used in netspoc main program to create ruleset for devices.

 - Aggregate with IPv4 IP 0.0.0.0/0 no longer matches IPv6 addresses.
   Aggregate with IPv6 IP ::/0 no longer matches IPv4 addresses.
 - Fixed split service from auto intf with IPv4, IPv6
 - Command 'export-netspoc' no longer strips comments and ";"
   from description in details of exported service.

6.077     2025-01-27 11:24:26+01:00 Europe/Berlin

 - Allow dual stack service to have protocols icmp and icmpv6 together.
   Generated rules automatically will use icmp for IPv4 and icmpv6 for IPv6.
 - Removed support for separate IPv6 topology in subdirectory 'ipv6/'.
 - Removed options '--ipv6', '-6'.
 - Removed support for NAT of IPv6 objects.
   After changing attributes of IPv6 addresses from 'ip' to 'ip6',
   supporting NAT for IPv6 would need to change NAT attributes
   to 'ip6', 'hidden6', 'identity6' as well.
   This seems to be superflous, since nobody uses NAT for IPv6.
 - Fixed duplicate hosts in print-group from wrong IPv4 network
   in IPv6 aggregate.
 - Fixed panic in print-group from wrong IPv4 network in IPv6 hosts
   generated by 'auto_ipv6_hosts'.
 - Fixed panic in export-netspoc with rule from IPv4 network to
   combined IPv4/v6 auto interface.
 - Changed 'format-netspoc' to print attributes "ip" and "ip6" in
   one line, if object has no more than 3 attributes.

6.076     2025-01-10 09:39:54+01:00 Europe/Berlin

 - Fixed dual stack bridge. IPv4 and IPv6 interfaces were accidently
   processed together.

6.075     2024-12-19 11:53:37+01:00 Europe/Berlin

 - Fix: Attribute 'policy_distribution_point' at dual stack router
   is only used for matching IPv4/IPv6 part of this router.
 - Fixed command 'export-netspoc'.
   Show IPv4 areas for combined zone. Must not accidently overwrite
   IPv4 areas by IPv6 areas, when combined zones have same name.

6.074     2024-12-16 11:52:23+01:00 Europe/Berlin

 - Fixed wrong error on duplicate interface from router in ipv6/ directory.

6.073     2024-12-16 10:16:33+01:00 Europe/Berlin

 - Added support for dual stack objects, having both, IPv4 and IPv6 addresses.
   This simplifies the modeling of a dual stack toplology.
   It is no longer neccessary to model a separate IPv4 and IPv6 topology.
   Rules between dual stack objects will generate ACls for IPv4 and IPv6.
   Rules between dual stack object and pure IPv4 object will silently ignore
   IPv6 address and generate only ACL for IPv4.
   New attributes have been introduced to define dual stack objects:
   - 'ip6' at network, host, interface and aggregate.
   - 'range6' at host.
   - 'unnumbered6' at network and interface.
   - 'negotiated6' at interface.
   Other changes resulting from use of dual stack objects:
   - If dual stack objects are used as border or inclusive border of an area,
     this defines two areas with identical name:
     one in IPv4 topology and one in IPv6 topology.
   - If dual stack objects are used as interfaces of an pathrestriction,
     this also defines two pathrestrictions in IPv4 and IPv6 topology.
     If the second pathrestriction has only one interface or only interfaces
     outside of a loop, it is silently ignored.
   - New attributes 'ipv4_only' and 'ipv6_only' may be used at service or area.
     This will enable only IPv4 or IPv6 part of dual stack objects.
   - The following attributes are applied only to IPv4 part
     if used in dual stack objects:
     'nat', 'bind_nat', 'subnet_of, 'hub', 'spoke'.
   - Command 'export-netspoc' has been changed to show IPv6 address of
     dual stack objects in attribute 'ip6'.
 - Added new attribute 'auto_ipv6_hosts' to automatically generate
   dual stack hosts from pure IPv4 hosts.
   It will generate IPv6 addresses for hosts
   by combining its IPv4 adress with the IPv6 address of its network.
   These attribute values are provided:
   - auto_ipv6_hosts = readable
     network: ip6 = 2001:db8:1:1::/64;
     hosts: ip = 172.17.1.48;
     ==> 2001:db8:1:1:172:17:1:48
   - auto_ipv6_hosts = binary
     network: ip6 = 2001:db8:1:1::/64;
     hosts: ip = 172.17.1.48;
     ==> 2001:db8:1:1::ac11:130
   - auto_ipv6_hosts = none
     No IPv6 address is generated
   This attribute can be used at network, area and host.
 - Fixed command 'cut-netspoc':
   Full path to management_instance is marked now.

6.072     2024-09-05 10:22:13+02:00 Europe/Berlin

 - Fixed command 'remove-from-netspoc'.
   A service is no longer removed accidently, if only some rules of
   a service become empty, but other rules remain.
 - Message "swap objects of 'user' and objects of rules" is shown
   even if no attribute multi_owner is given.
 - Fixed check for NAT network being subnet of some other network.
   If multiple networks have dynamic NAT to same address,
   now a warning is given for each network.
 - Fixed check for useless attribute 'subnet_of'
   where supernet has attribute 'has_subnets'.

6.071     2024-07-29 16:47:58+02:00 Europe/Berlin

 - Fixed generated ACL at router R with managed=local for rule where

go/pkg/pass1/print-service.go
package pass1

import (
	"fmt"
	"io"
	"maps"
	"slices"
	"strconv"
	"strings"

	"github.com/hknutzen/Netspoc/go/pkg/conf"
	"github.com/hknutzen/Netspoc/go/pkg/oslink"
	"github.com/spf13/pflag"
)

func prtInfo(srcRange, prt *proto) string {
	ptype := prt.proto
	desc := ptype
	switch ptype {
	case "tcp", "udp":
		portCode := func(rangeObj *proto) string {
			v1, v2 := rangeObj.ports[0], rangeObj.ports[1]
			if v1 == v2 {
				return strconv.Itoa(v1)
			} else if v1 == 1 && v2 == 65535 {
				return ""
			} else {
				return strconv.Itoa(v1) + "-" + strconv.Itoa(v2)
			}
		}
		var sport string
		if srcRange != nil {
			sport = portCode(srcRange)
		}
		dport := portCode(prt)
		if sport != "" {
			desc += " " + sport + ":" + dport
		} else if dport != "" {
			desc += " " + dport
		}
	case "icmp", "icmpv6":
		if t := prt.icmpType; t != -1 {
			s := strconv.Itoa(t)
			if c := prt.icmpCode; c != -1 {
				desc += " " + s + "/" + strconv.Itoa(c)
			} else {
				desc += " " + s
			}
		}
	}
	return desc
}

func (c *spoc) printService(
	stdout io.Writer,
	path string,
	srvNames []string, natNet string, showName bool, showIP bool) {

	c.readNetspoc(path)
	c.setZone()
	c.setPath()
	c.distributeNatInfo()
	c.stopOnErr()

	if !(showIP || showName) {
		showIP = true
		showName = false
	}

	// Find network for resolving NAT addresses or use empty map.
	var natMap natMap
	if natNet != "" {
		natNet = strings.TrimPrefix(natNet, "network:")
		if net := c.symTable.network[natNet]; net != nil {
			natMap = net.zone.natDomain.natMap
		} else {
			c.abort("Unknown network:%s of option '--nat'", natNet)
		}
	}

	sRules := c.normalizeServices()
	permitRules, denyRules := c.convertHostsInRules(sRules)
	c.stopOnErr()

	nameMap := make(map[string]bool)
	for _, name := range srvNames {
		name = strings.TrimPrefix(name, "service:")
		if _, found := c.symTable.service[name]; !found {
			c.abort("Unknown service:%s", name)
		}
		nameMap[name] = true
	}

	// Collect expanded rules.
	type rule struct {
		deny     bool
		src      someObj
		dst      someObj
		srcRange *proto
		prt      *proto
	}
	s2rules := make(map[string][]rule)
	// Remove duplicates resulting from aggregates of zone cluster.
	uniq := func(l []someObj) []someObj {
		var prev string
		j := 0
		for _, ob := range l {
			if n, ok := ob.(*network); ok {
				if n.name == prev {
					continue
				}
				prev = n.name
			}
			l[j] = ob
			j++
		}
		return l[:j]
	}
	collect := func(rules ruleList) {
		for _, r := range rules {
			sName := r.rule.service.name
			sName = strings.TrimPrefix(sName, "service:")
			if len(nameMap) != 0 && !nameMap[sName] {
				continue
			}
			sList := uniq(r.src)
			dList := uniq(r.dst)
			for _, src := range sList {
				for _, dst := range dList {
					for _, prt := range r.prt {
						s2rules[sName] = append(
							s2rules[sName],
							rule{
								deny:     r.deny,
								src:      src,
								dst:      dst,
								srcRange: r.srcRange,
								prt:      prt})
					}
				}
			}
		}
	}
	collect(denyRules)
	collect(permitRules)

	objInfo := func(obj someObj) string {
		var result stringList
		if showIP {
			result.push(prefixCode(obj.address(natMap)))
		}
		if showName {
			result.push(obj.String())
		}
		return strings.Join(result, " ")
	}

	for _, name := range slices.Sorted(maps.Keys(s2rules)) {
		for _, r := range s2rules[name] {
			action := "permit"
			if r.deny {
				action = "deny"
			}
			srcInfo := objInfo(r.src)
			dstInfo := objInfo(r.dst)
			prtInfo := prtInfo(r.srcRange, r.prt)
			fmt.Fprintf(stdout, "%s:%s %s %s %s\n",
				name, action, srcInfo, dstInfo, prtInfo)
		}
	}
}

func PrintServiceMain(d oslink.Data) int {
	fs := pflag.NewFlagSet(d.Args[0], pflag.ContinueOnError)

	// Setup custom usage function.
	fs.Usage = func() {
		fmt.Fprintf(d.Stderr,
			"Usage: %s [options] FILE|DIR [SERVICE-NAME ...]\n%s",
			d.Args[0], fs.FlagUsages())
	}

	// Command line flags
	quiet := fs.BoolP("quiet", "q", false, "Don't print progress messages")

	nat := fs.String("nat", "",
		"Use network:name as reference when resolving IP address")
	name := fs.BoolP("name", "n", false, "Show only name of elements")
	ip := fs.BoolP("ip", "i", false, "Show only IP address of elements")
	if err := fs.Parse(d.Args[1:]); err != nil {
		if err == pflag.ErrHelp {
			return 1
		}
		fmt.Fprintf(d.Stderr, "Error: %s\n", err)
		fs.Usage()
		return 1
	}

	// Argument processing
	args := fs.Args()
	if len(args) == 0 {
		fs.Usage()
		return 1
	}
	path := args[0]
	names := args[1:]

	cnf := conf.ConfigFromFile(path)
	cnf.Quiet = *quiet
	return toplevelSpoc(d, cnf, func(c *spoc) {
		c.printService(d.Stdout, path, names, *nat, *name, *ip)
	})
}
